File: algorithms/data-structures/skip-list/swift/SkipList.swift
private let MAX_LVL = 16

private class SkipNode {
    var key: Int
    var forward: [SkipNode?]
    init(_ key: Int, _ level: Int) {
        self.key = key
        self.forward = [SkipNode?](repeating: nil, count: level + 1)
    }
}

func skipList(_ arr: [Int]) -> [Int] {
    let header = SkipNode(Int.min, MAX_LVL)
    var level = 0

    for val in arr {
        var update = [SkipNode?](repeating: nil, count: MAX_LVL + 1)
        var current = header
        for i in stride(from: level, through: 0, by: -1) {
            while let fwd = current.forward[i], fwd.key < val {
                current = fwd
            }
            update[i] = current
        }
        if let next = current.forward[0], next.key == val { continue }

        var newLevel = 0
        while Bool.random() && newLevel < MAX_LVL { newLevel += 1 }
        if newLevel > level {
            for i in (level + 1)...newLevel { update[i] = header }
            level = newLevel
        }
        let newNode = SkipNode(val, newLevel)
        for i in 0...newLevel {
            newNode.forward[i] = update[i]!.forward[i]
            update[i]!.forward[i] = newNode
        }
    }

    var result: [Int] = []
    var node = header.forward[0]
    while let n = node {
        result.append(n.key)
        node = n.forward[0]
    }
    return result
}

File: algorithms/data-structures/skip-list/python/skip_list.py
import random

def skip_list(arr: list[int]) -> list[int]:
    MAX_LEVEL = 16

    class Node:
        def __init__(self, key, level):
            self.key = key
            self.forward = [None] * (level + 1)

    level = 0
    header = Node(-1, MAX_LEVEL)

    def random_level():
        lvl = 0
        while random.random() < 0.5 and lvl < MAX_LEVEL:
            lvl += 1
        return lvl

    def insert(key):
        nonlocal level
        update = [None] * (MAX_LEVEL + 1)
        current = header
        for i in range(level, -1, -1):
            while current.forward[i] and current.forward[i].key < key:
                current = current.forward[i]
            update[i] = current
        current = current.forward[0]
        if current and current.key == key:
            return
        new_level = random_level()
        if new_level > level:
            for i in range(level + 1, new_level + 1):
                update[i] = header
            level = new_level
        new_node = Node(key, new_level)
        for i in range(new_level + 1):
            new_node.forward[i] = update[i].forward[i]
            update[i].forward[i] = new_node

    for val in arr:
        insert(val)

    result = []
    node = header.forward[0]
    while node:
        result.append(node.key)
        node = node.forward[0]
    return result
